lips/lip-10.md

```diff
@@ -0,0 +1,190 @@
+
+## Backend flow
+In order to support peer-2-peer payments, VASPs need to implement two Libra ID endpoints:
+* **Info endpoint** for querying optional information about the receiver. VASPs can provide this endpoint to improve UX and security of payment, but they are not obligated to - VASPs may choose not to share any information about the user via this endpoint, including whether the user exists or not
+* **Payment endpoint** to perform actual money transfer.
+
+Common considerations:
+
+* For both endpoints, all communications are signed via the same mechanism as for other off-chain communications, see the [Payloads section of lip-1](https://github.com/libra/lip/blob/master/lips/lip-1.mdx#payloads) for details. As such, all vasp-2-vasp communications are authenticated.
+
+* All Libra ID endpoints are nested under `pay.<version>` URI path:
+
+     `https://<hostname>:<port>/pay.<protocol_version>/<endpoint>`
+
+* All Libra ID endpoints use the same `base_url` and same `compliance_public_key` for authentication as other off-chain APIs.
+
+* All Libra ID endpoints are HTTP POST endpoints, they accept JSON-encoded requests and return JSON-encoded responses. Other encoding methods can be considered in the future, `Content-Type` and `Accept` HTTP headers can be used to define request/response formats.
+
+## Info endpoint
+
+```
+POST https://<hostname>:<port>/pay.<protocol_version>/info
+```
+
+**Info endpoint** is used by the **Sender** to request information about the **Receiver** before submitting an actual payment.
+
+Info endpoint takes both sender and receiver Libra IDs as arguments. Receiver VASP might decide to share more(or less) information based on sender ID, for example if the wallet supports some kind of contact list, they might decide to share more information if the sender is in the contact list of the receiver.
+
+**Request:**
+```json
+{
+   "sender_id": "<Sender Libra ID>",
+   "receiver_id": "<Receiver Libra ID>"
+}
+```
+
+**Response:**
+```json
+{
+   "receiver": <User Object>
+}
+```
+
+* Receiver Libra ID domain should belong to the Receiver VASP. Otherwise, Receiver VASP should return the `400` error code.
+* Sender Libra ID should have a Libra ID domain that belongs to Sender VASP. If receiver VASP makes a decision on how much information to share, based on the sender Libra ID, it should enforce this property and return 400 error code, if VASP tries to impersonate a user from Libra ID domain that it does not own.
+
+## Pay endpoint
+
+```
+POST https://<hostname>:<port>/pay.<protocol_version>/pay
+```
+
+**Pay endpoint** is used when a user is willing to make a money transfer.
+
+**Request:**
+```json
+{
+   "sender": <User Object>,
+   "receiver_id": "<Receiver_Libra_ID>",
+   "amount": "<amount>",
+   "currency": "<currency>"
+}
+```
+
+**Response:**
+```json
+{
+   "receiver_address": "<Receiver VASP on-chain account address>",
+   "reference_id": "<Payment_Reference_ID>",
```

Comment:
Hmmm, as I look at it more, I do wonder if we actually need two endpoints.  It feels like this is really just one endpoint that has a flag for if you want a reference ID or not.  Otherwise, they are effectively the same thing

Reply:
To me two separate endpoints makes more sense. Yes they share some fields (like there is sender/receiver), but there are also many fields that are not shared (info endpoint don't have amounts/coin in request, and does not return reference/account address)
They also have different semantics - info endpoint is 'read' endpoint, while 'pay' is a 'write' endpoint as it needs to create an entry in the database

Reply:
This is an open write API, this is dangerous. Folks will mess this up and have vulnerabilities.